Ingest

aumayr opened the pull request:
This PR adds initial `ingest`-functionality described in #436 to Fava:

- Upload a (CSV-) file
- Parse the file and display multiple transaction forms
- Save the transactions to the beancount file

This PR includes a sample bank CSV importer, which works with the following files:
- CSV: [testfile.csv.zip](https://github.com/beancount/fava/files/775294/testfile.csv.zip) (This is similar to what one of my banks (sorry for the German language in the file) let's me download on their website)
- Beancount: [example.beancount.zip](https://github.com/beancount/fava/files/775296/example.beancount.zip)

You can enable the feature by declaring "ingesters" as a Fava option like in the example beancount file:

```ruby
2016-04-14 custom "fava-option" "ingesters" "fava.ingest.sample_bank_csv_ingester"
```

This is what the UI looks like (with the files from above, 2 transactions):

![screenshot](https://cloud.githubusercontent.com/assets/5135450/22949501/c8d836c2-f302-11e6-8a5d-6cb5278d4f94.png)

Feedback (and help) is more than welcome!

yagebu commented:
This looks nice! Will have a more in-depth look at the code later today or tomorrow. Some comments after a quick glance:

- We shouldn't be duplicating the functionality in `beancount.ingest` - Beancount already has a standardised way of writing importers, which we should use (and remove `FavaIngestBase `).
- If the page is called "Import" so should the URL, API endpoint etc. - having both "ingest" and "import" with the same meaning in various places is bad.

aumayr commented:
- Renamed everyting to `ingest` instead of `import`
- As of 4aa3ec5 it uses the Beancount ingest mechanism. The usage of importers can now be configured in this way:

```ruby
2016-04-14 custom "fava-option" "ingest-config" "testimporter.config"
2016-04-14 custom "fava-option" "ingest-dirs" "/Users/dominik/Downloads"
```

Updated TODO-List:

### TODO

- [ ] "Intelligent" way to suggest accounts, etc. (like discussed in #436)
- [ ] Support for more types of entries (`balance` and `note` seem important)
- [ ] Tests

aumayr commented:
As a next step for this PR, I want to implement a file upload, and then subsequently use this file (instead of the config-specified directory and the files therein) for the ingest-steps.

@ Do you have any advice how to store the file temporarily to feed it to the beancount ingest functions?

aumayr commented:
As a next step for this PR, I want to implement a file upload, and then subsequently use this file (instead of the config-specified directory and the files therein) for the ingest-steps.

@yagebu Do you have any advice how to store the file temporarily to feed it to the beancount ingest functions?

corani commented:
I'm guessing you could probably generate a temporary file using e.g.
NamedTemporaryFile, feed it to beancount ingress and afterwards unlink it.
But thinking about it, it might make sense to *not* use a temporary file,
but store the file as a linked document. At least, I like to do this with
my bank statements.

On Tue, Mar 28, 2017 at 2:20 PM, Dominik Aumayr <notifications@github.com>
wrote:

> As a next step for this PR, I want to implement a file upload, and then
> subsequently use this file (instead of the config-specified directory and
> the files therein) for the ingest-steps.
>
> @yagebu <https://github.com/yagebu> Do you have any advice how to store
> the file temporarily to feed it to the beancount ingest functions?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/beancount/fava/pull/503#issuecomment-289674675>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdWB0FTdq-mM7_P36kfr_0Zi2JbEWKZks5rqKbFgaJpZM4MBAKJ>
> .
>



-- 
Best regards,
Daniël Bos


Your government is reading your email. Slow them down with encryption.

My public key: http://goo.gl/gms497 (4096 bit RSA, id EF2D5D91)
Fingerprint  : D8D0 9FBE F075 F709 7B52  2F73 326C 2123 EF2D 5D91

yagebu commented:
> Do you have any advice how to store the file temporarily to feed it to the beancount ingest functions?

If it only needs to be passed to some Python functions, we can also just keep it in memory, a tempfile doesn't add much then and is just a lot more hassle. Figuring out a place to store it permanently might also be nice though.

aumayr commented:
> If it only needs to be passed to some Python functions, we can also just keep it in memory, a tempfile doesn't add much then and is just a lot more hassle. Figuring out a place to store it permanently might also be nice though.

Right now I solved with a temporary directory, as I don't know how to store it in memory over multiple requests (There are two steps required: Parse the file to figure out which importer to use, as there may be more than one, and then parse it again with the user-selected importer to get the entries).

yagebu commented:
> Right now I solved with a temporary directory, as I don't know how to store it in memory over multiple requests

But then you have to store the temp file path somewhere over multiple requests, right? Just store the file there instead.

aumayr commented:
> But then you have to store the temp file path somewhere over multiple requests, right? Just store the file there instead.

`session` is stored in a cookie on the client, therefore there is a limit on how much you can store there. That's why storing the paths to files is OK, but the whole files is not.

yagebu commented:
> session is stored in a cookie on the client, therefore there is a limit on how much you can store there. That's why storing the paths to files is OK, but the whole files is not.

I see. Then store it in a dict and store the dict key in the session

aumayr commented:
> I see. Then store it in a dict and store the dict key in the session

You mean a dict in the application (global) context?

yagebu commented:
Maybe on the active `FavaLedger` instance is more appropriate but global would be fine too.

yagebu commented on fava/core/fava_options.py:
```diff
@@ -54,11 +56,13 @@
     'journal-show',
     'journal-show-document',
     'journal-show-transaction',
+    'ingest-dirs'
```
add trailing comma

yagebu commented on fava/core/fava_options.py:
```diff
@@ -54,11 +56,13 @@
 STR_OPTS = [
     'language',
     'interval',
+    'ingest-config'
```
add trailing comma

yagebu commented on fava/application.py:
```diff
@@ -290,6 +294,98 @@ def help_page(page_slug='_index'):
                            help_html=render_template_string(html))
 
 
+def _load_ingest_config(configpath):
```
All this logic shouldn't be in `fava.application` but rather in `fava.core.import`

yagebu commented on fava/application.py:
```diff
@@ -290,6 +294,98 @@ def help_page(page_slug='_index'):
+    return mod['CONFIG']
+
+
+def _identify_importers(config, file_or_dir):
```
Same here.

yagebu commented on fava/json_api.py:
```diff
@@ -95,28 +95,79 @@ def add_document():
+    g.ledger.file.insert_entry(transaction)
+
+
+def _add_balance(json):
```
Now that this is getting more complicated, I feel like this should live in `core.file` and have some tests.

aumayr replied:
Maybe we should generally rethink the architecture around adding entries via Fava:

- Support for all types of entries would be awesome (not just `transaction`, `balance` and `note`)
- Support for all types in the "+"-popup
- Formsets that can be toggled to just plain `textfield`s (for more granular control)
- JS that is more generic
- Stricter parsing of the entries from JSON
- Move everything to `core.file`

Maybe tackle this in a separate PR?

yagebu replied:
> Maybe tackle this in a separate PR?

👍  the functionality needed for this PR could simplify reviewing it (and with tests, it could be merged even if there's no frontend using it yet)

aumayr commented:
I adressed some of the points raised above @yagebu - thanks for the input! 
The code should be cleaner now, as it is in it's own `FavaModule`.

What I am struggling with is the rebasing so the merge can be done without conflicts. How do I go about this the right way? I don't want to screw up `git` again 🙄

yagebu commented:
> What I am struggling with is the rebasing so the merge can be done without conflicts. How do I go about this the right way?

Probably easiest to merge master into this one and then squash everything into one commit. (With stuff like the CSS I guess git won't be able to help you since too much changed so you'll just have to copy it over by hand)

aumayr commented:
This PR is complete. (I can't figure out why one of the Travis tests fails 😕)

Only four points remain, which can be implemented after the PR is merged:

- Figure out (and integrate) the duplicate-entry-detection-mechanism by `bean-extract`. For me, it never recognises an entry as duplicate. If duplicates are recognised, mark them in the Ingest-UI as "duplicate". 
- Add `bean-file` functionality.
- The Ingest-UI is slow for >50 entries. 
- Add example code for how to implement an importer and configure Fava to use it. This should also contain a blueprint for a best-practice importer.

yagebu commented:
Can you squash everything into one commit?

> I can't figure out why one of the Travis tests fails

The lint is failing in `core.ingest` and `json_api` as the log shows.

yagebu commented on fava/application.py:
```diff
@@ -294,6 +301,42 @@ def help_page(page_slug='_index'):
+    """Identify files to ingest."""
+    sid = session['sid'] if 'sid' in session else None
+    return render_template('ingest.html',
+                           identified=g.ledger.ingest.dirs_importers(),
```
This should be done in the template (same in line below)

yagebu commented on fava/core/ingest.py:
```diff
@@ -0,0 +1,92 @@
+IngestError = namedtuple('IngestError', 'source message entry')
+
+
+# pylint: disable=too-few-public-methods
```
Remove

yagebu commented on fava/core/ingest.py:
```diff
@@ -0,0 +1,92 @@
+
+        return identified
+
+    def add_upload(self, session_id, file):
```
Needs docstring. Same for the other functions.

yagebu commented on fava/core/ingest.py:
```diff
@@ -0,0 +1,92 @@
+        super().__init__(ledger)
+        self.config = None
+        self.ingest_dirs = []
+        self.uploaded_files = defaultdict(list)
```
This should just be list (not a dict of lists). So far all application state is either global or relative to a ledger and so should this be IMHO. Using sessions for this just seems to lead to unnecessary complication.

yagebu commented on fava/core/ingest.py:
```diff
@@ -0,0 +1,92 @@
+
+
+# pylint: disable=too-few-public-methods
+class IngestModule(FavaModule):
```
This Module badly needs tests.

yagebu commented on fava/templates/ingest.html:
```diff
@@ -0,0 +1,123 @@
+            </div>
+        </form>
+    {% else %}
+        {% set identified = ledger.ingest.dirs_importers() %}
```
No need to define a variable here, just use directly.

yagebu commented on fava/templates/ingest.html:
```diff
@@ -0,0 +1,123 @@
+                                <td>{{ importer.name }}</td>
+                                <td>{{ account_macros.account_name(importer.account) }}</td>
+                                <td>
+                                    <button class="a inactive" title="{{ _('Delete') }} {{ file.filename }}" href="{{ url_for('ingest_delete') }}?filename={{ file.filename|e }}">{{ _('Delete') }}</button>
```
Don't build URLs by hand, use `url_for`

yagebu replied:
Also, this is just a link, not a button so use `a` not `button`

yagebu commented on fava/templates/ingest.html:
```diff
@@ -0,0 +1,123 @@
+        {% endfor %}
+        <h3>{{ _('Upload file') }}</h3>
+        {% set uploaded = ledger.ingest.uploads_importers() %}
+        {% if uploaded|length > 0 %}
```
`if uploaded` is more pythonic

yagebu commented on fava/static/javascript/ingest.js:
```diff
@@ -0,0 +1,120 @@
+  const form = $('.ingest-extract');
+  const jsonData = { entries: [] };
+
+  form.querySelectorAll('.ingest-row.import .transaction-form').forEach((transaction) => {
```
This code is very repetitive. Can't we iterate over all `.ingest-row.import`, for each of them iterate over all `input`s and then set `entryData[input.name] = input.value`. The "type" can also be determined from the HTML, so we'd only need to special-case postings, but otherwise we could handle transactions, balances, and notes in one go.

yagebu commented on fava/application.py:
```diff
@@ -294,6 +295,32 @@ def help_page(page_slug='_index'):
+    """Extract entries to ingest."""
+    filename = request.args.get('filename')
+    importer_name = request.args.get('importer')
+    entries = g.ledger.ingest.extract(filename, importer_name)
```
Better do this in the template.

yagebu commented on fava/core/ingest.py:
```diff
@@ -0,0 +1,94 @@
+
+    def add_upload(self, file):
+        """Store the file in a temporary directory."""
+        new_file = os.path.join(tempfile.mkdtemp(), file.filename)
```
Better just store it in memory. Otherwise we'll have to handle cleanup for the temp files which is a hassle.

aumayr replied:
The problem is that the beancount functions only work on paths:

https://github.com/beancount/fava/pull/503/files#diff-88568880c80b13bcb45b2e04c7253947R44
https://github.com/beancount/fava/pull/503/files#diff-88568880c80b13bcb45b2e04c7253947R84

Any suggestions on how to work around that while keeping the file in memory?

aumayr replied:
Ahh, just found the answer myself: Use a tempfile just for this one (or two) line(s), and create it every time it is used. This way the beancount-functions have a real file to work with.

yagebu replied:
Then I think it's better to store it in one of the ingest-dirs. If you upload a document it naturally goes to the document folder, not some temp folder that just disappears when you quit Fava, and files to import seem quite similar to documents.

yagebu commented on fava/core/ingest.py:
```diff
@@ -0,0 +1,94 @@
+        new_entries, duplicate_entries = extract.extract_from_file(
+            filepath,
+            importer,
+            existing_entries=self.ledger.entries,
```
`all_entries`

yagebu commented:
I'd be for removing the upload functionality from this PR. It's far from ready and the core "ingest" functionality is quite independent of it. It can easily be added at a later stage and then it wouldn't slow down the merge of this PR.

yagebu commented:
> Renamed everyting to ingest instead of import.

I strongly believe it should be the other way round. "Import" is the established term for this sort of functionality and using a neologism in the user interface is a very bad idea IMHO.

yagebu commented:
@aumayr: Could you write a test for the ingest functionality using your example importer and file?

yagebu commented:
Will merge over some of the changes into master by hand to reduce the diff here a bit (since the history of this PR is probably FUBAR :D )

yagebu commented on tests/data/import.csv:
```diff
@@ -0,0 +1,3 @@
+IBAN;Auszugsnummer;Buchungsdatum;Valutadatum;Umsatzzeit;Zahlungsreferenz;Waehrung;Betrag;Buchungstext;Umsatztext
+ATXYZ123400000;1;2017-02-14;2017-02-14;2017-02-14-11.57.44.235044;"";EUR;-100,00;"Bankomat";" BANKOMAT  00000483 K2 UM 11:56"
```
@aumayr: Do you have some lines that generate Notes?

aumayr replied:
What exactly do you mean? Lines in the CSV-file that would correspond to/generate `note`s in a beancount ledger?

yagebu replied:
Yes. Would be good if the UI for all kinds of entries could be tested with this file.

aumayr replied:
I do not have such lines in the files I personally use, but we could come up with a simple example, like so:

```csv
IBAN;Auszugsnummer;Buchungsdatum;Valutadatum;Umsatzzeit;Zahlungsreferenz;Waehrung;Betrag;Buchungstext;Umsatztext;Saldo
ATXYZ123400000;1;2017-02-14;2017-02-14;2017-02-14-11.57.44.235044;"";EUR;-100,00;"Bankomat";" BANKOMAT  00000483 K2 UM 11:56";4.567,89
;;;;2017-02-14;;;;Kundenhinweis: Zinssatz auf 0,15% geändert;;
```

This example does two things:
- Add a column at the end for a balance, to be used for a final `balance` entry
- Add a line which can be parsed to a `note`

aumayr commented:
Any further comments or TODOs?

yagebu commented:
Tests for the json_api part would be good.

aumayr commented:
> Tests for the json_api part would be good.

Can we implement the tests after this is merged?

yagebu commented:
> Can we implement the tests after this is merged?

That'd be bad practice IMHO. (Sorry for only replying now, had a busy week).

I've used this feature for the first time for a real import. It worked quite well but the entries were inserted in the wrong order. I guess we need to sort them before importing them.

aumayr commented:
🎉

Thanks @yagebu for the help!
